pkg/ccl/backupccl/restore_job.go

```diff
@@ -2234,24 +2234,36 @@ func (r *restoreResumer) dropDescriptors(
 
 	// Delete any schema descriptors that this restore created. Also collect the
 	// descriptors so we can update their parent databases later.
-	dbsWithDeletedSchemas := make(map[descpb.ID][]catalog.SchemaDescriptor)
+	dbsWithDeletedSchemas := make(map[descpb.ID][]catalog.Descriptor)
 	for _, schemaDesc := range details.SchemaDescs {
-		sc := schemadesc.NewBuilder(schemaDesc).BuildImmutableSchema()
 		// We need to ignore descriptors we just added since we haven't committed the txn that deletes these.
-		isSchemaEmpty, err := isSchemaEmpty(ctx, txn, sc.GetID(), allDescs, ignoredChildDescIDs)
+		isSchemaEmpty, err := isSchemaEmpty(ctx, txn, schemaDesc.GetID(), allDescs, ignoredChildDescIDs)
 		if err != nil {
-			return errors.Wrapf(err, "checking if schema %s is empty during restore cleanup", sc.GetName())
+			return errors.Wrapf(err, "checking if schema %s is empty during restore cleanup", schemaDesc.GetName())
 		}
 
 		if !isSchemaEmpty {
-			log.Warningf(ctx, "preserving schema %s on restore failure because it contains new child objects", sc.GetName())
+			log.Warningf(ctx, "preserving schema %s on restore failure because it contains new child objects", schemaDesc.GetName())
 			continue
 		}
 
-		b.Del(catalogkeys.EncodeNameKey(codec, sc))
-		b.Del(catalogkeys.MakeDescMetadataKey(codec, sc.GetID()))
-		descsCol.AddDeletedDescriptor(sc)
-		dbsWithDeletedSchemas[sc.GetParentID()] = append(dbsWithDeletedSchemas[sc.GetParentID()], sc)
+		mutSchema, err := descsCol.GetMutableDescriptorByID(ctx, schemaDesc.GetID(), txn)
+		if err != nil {
+			return err
+		}
+
+		// Mark schema as dropped and add uncommitted version to pass pre-txn
+		// descriptor validation.
+		mutSchema.SetDropped()
+		mutSchema.MaybeIncrementVersion()
+		if err := descsCol.AddUncommittedDescriptor(mutSchema); err != nil {
+			return err
+		}
+
+		b.Del(catalogkeys.EncodeNameKey(codec, mutSchema))
+		b.Del(catalogkeys.MakeDescMetadataKey(codec, mutSchema.GetID()))
+		descsCol.AddDeletedDescriptor(mutSchema)
+		dbsWithDeletedSchemas[mutSchema.GetParentID()] = append(dbsWithDeletedSchemas[mutSchema.GetParentID()], mutSchema)
```

Comment:
I think we should store to an immutable schema descriptor in this map. Even though it doesn't change much it prevents us from unintentionally mutating the descriptor wherever we use the values in this map.

Reply:
No worry in this case because what we are adding into `dbsWithDeletedSchemas` map is catalog.Descriptor which only has immutable APIs.